feat(rule-engine): introduce AML_CHECK_V2 operator for enhanced AML checks

tomer-shvadron opened the pull request:
- Add AML_CHECK_V2 operator to validate adverse media, PEP, sanctions, and fitness probity
- Implement rule evaluation to return results based on the presence of AML hits
- Update TypeScript types and schemas to support new operator logic and helpers


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new AML_CHECK_V2 operator for advanced Anti-Money Laundering checks, supporting evaluation across multiple AML categories and entity roles.
  - Added support for helper functions (e.g., retrieving end user data) during rule evaluation, enabling more flexible and dynamic rule processing.

- **Bug Fixes**
  - Improved error handling and validation in AML-related operators.

- **Tests**
  - Added comprehensive tests for AML_CHECK_V2 and updated existing tests to utilize the new helpers context.

- **Chores**
  - Updated interfaces and service methods to support and propagate the helpers context throughout rule engine workflows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tomer-shvadron committed:
feat(rule-engine): introduce AML_CHECK_V2 operator for enhanced AML checks

- Add AML_CHECK_V2 operator to validate adverse media, PEP, sanctions, and fitness probity
- Implement rule evaluation to return results based on the presence of AML hits
- Update TypeScript types and schemas to support new operator logic and helpers

tomer-shvadron committed:
refactor(aml-v2-check): improve null handling for ubos and directors IDs

- Replace implicit null checks with optional chaining and default values
- Filter out any falsy IDs before fetching end users
- Enhance code readability and maintainability by streamlining mapping operations

tomer-shvadron committed:
feat(tests): add comprehensive unit tests for AML v2 rules

- Introduce new unit tests for AML v2, verifying multiple rule outcomes
- Implement helper function for fetching user details in test scenarios
- Refactor existing company sanctions test cases to utilize helper function

tomer-shvadron committed:
refactor(tests): consolidate helper functions into data-helper module

- Move getEndUserById function to data-helper for better organization
- Remove redundant helper definitions in individual test files
- Ensure that all test modules now import the shared helpers from data-helper

tomer-shvadron committed:
refactor(types): make helpers property optional in WorkflowOptions and WorkflowRunnerArgs

- Update WorkflowOptions and WorkflowRunnerArgs interfaces to allow helpers to be optional
- Adjust related types for consistency and flexibility in workflow handling

alonp99 commented on packages/common/src/rule-engine/operators/aml-v2-check.ts:
```diff
@@ -0,0 +1,92 @@
+        .map(endUserId => helpers.getEndUserById(endUserId)),
+    );
+
+    if (isEmpty(endUsers)) {
```
Consider checking before the db call if there are and ids to check, and just return.

if there are some ID's but nothing returned from the database, something is wrong and you should trigger sentry or at least a log

alonp99 commented on packages/common/src/rule-engine/operators/aml-v2-check.ts:
```diff
@@ -0,0 +1,92 @@
+      return false;
+    }
+
+    const hits: Array<z.infer<typeof EndUserAmlHitsSchema>> = endUsers
```
Since we know all won't be on the user in the future, it's worth mentioning it in a comment here to help future dev thiking why this rule is not working or where the refactor should happen in this rule

alonp99 commented on packages/common/src/rule-engine/operators/aml-v2-check.ts:
```diff
@@ -0,0 +1,92 @@
+      const conditionResult = amlOperator.conditionValueSchema?.safeParse(conditionValue.value);
+
+      if ((conditionResult && !conditionResult.success) || !conditionResult?.data) {
+        return false; // TODO: throw explicit error
```
^

alonp99 commented on packages/common/src/rule-engine/operators/enums.ts:
```diff
@@ -12,6 +12,7 @@ export const OPERATION = {
   IN_CASE_INSENSITIVE: 'IN_CASE_INSENSITIVE',
   NOT_IN: 'NOT_IN',
   AML_CHECK: 'AML_CHECK',
+  AML_CHECK_V2: 'AML_CHECK_V2',
```
Do we need to migrate current v1 usage to v2?

tomer-shvadron replied:
Will be done in BAL-4113

alonp99 committed:
Merge branch 'dev' into bal-3872-add-risk-rules-for-directors-and-main-representative

tomer-shvadron committed:
fix(aml-check): improve end user validation logic and error handling

- Refactor end user ID collection to handle empty cases correctly
- Throw validation error with detailed message if no end users found
- Adjust validation condition for more explicit failure handling
